wandb/sdk/launch/runner/kubernetes_runner.py

```diff
@@ -99,6 +108,110 @@ def __init__(
         self._fail_count = 0
         self.pod_names = pod_names
         self.secret = secret
+        self._status_lock = Lock()
+        self._status = Status("running")
+        self._watch_job_thread = Thread(target=self._watch_job, daemon=True)
+        self._watch_pod_thread = Thread(
+            target=self._watch_pod, args=(pod_names[0],), daemon=True
+        )
+
+    def _set_status(self, status: Status) -> None:
+        """Set the run status."""
+        with self._status_lock:
+            self._status = status
+
+    def _watch_pod(self, pod_name: str) -> None:
+        """Handle event stream from rank 0 pod."""
+        w = watch.Watch()
+        try:
+            for event in w.stream(
+                self.core_api.list_namespaced_pod,
+                namespace=self.namespace,
+                field_selector=f"metadata.name={pod_name}",
+            ):
+                self._handle_pod_event(event)
+        except ApiException as e:
+            raise LaunchError(
+                "Exception when calling CoreV1Api->read_namespaced_pod: %s\n" % e
+            )
+
+    def _handle_pod_event(self, event: Dict[str, Any]) -> None:
+        """Update status according to an event on the rank 0 pod.
+
+        Arguments:
+            event: Kubernetes event. This is a dict that maps the string
+                "type" to the type of event and the string "object" to the
+                object that the event is about.
+
+        Returns:
+            None.
+
+        Raises:
+            LaunchError: If the pod fails to start after MAX_KUBERNETES_RETRIES.
+        """
+        pod = event.get("object")
+        if pod is None:
+            return
+        if hasattr(pod.status, "conditions") and pod.status.conditions is not None:
+            for condition in pod.status.conditions:
+                if condition.type == "DisruptionTarget" and condition.reason in [
+                    "EvictionByEvictionAPI",
+                    "PreemptionByScheduler",
+                    "TerminationByKubelet",
+                ]:
+                    self._set_status(Status("preempted"))
+        if pod.status.phase in ["Pending", "Unknown"]:
+            now = time.time()
+            if self._fail_count == 0:
+                self._fail_first_msg_time = now
+                self._fail_last_msg_time = 0.0
+            self._fail_count += 1
+            if now - self._fail_last_msg_time > FAIL_MESSAGE_INTERVAL:
+                wandb.termlog(
+                    f"{LOG_PREFIX}Pod has not started yet for job: {self.name}. Will wait up to {round(10 - (now - self._fail_first_msg_time)/60)} minutes."
+                )
+                self._fail_last_msg_time = now
+            if self._fail_count > MAX_KUBERNETES_RETRIES:
+                raise LaunchError(f"Failed to start job {self.name}")
```

Comment:
seems like we dont care about pod events after it is running, so we should probably stop the pod watch after it gets into a running state

Reply:
IMO, we should ensure the pod is in a terminal state before we release the agent thread.

Reply:
If job is failed/succeded then it's guaranteed that pods are done